<div style="border:solid Chocolate 2px; padding: 40px">

<b>Привет!👋</b>

Меня зовут Роман Абрамов, я буду ревьюером твоего проекта. Ты можешь обращаться ко мне на "ты"😏 Надеюсь, тебя также не смутит, если я буду обращаться к тебе на "ты", но если это неудобно, обязательно скажи об этом!

Пожалуйста, не удаляй мои комментарии, они будут особенно полезны для нашей работы в случае повторной проверки проекта.

Ты также можешь реагировать на мои комментарии своими, выделяя их цветами и наиболее понравившимся тебе способом оформления, но явно  отличающимся от моих комментариев. Это нужно, чтобы не создавалась путаница😉

Ты можешь найти мои комментарии, обозначенные <font color='green'>зеленым</font>, <font color='gold'>желтым</font> и <font color='red'>красным</font> цветами, например:

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> В случае, если решение на отдельном шаге является полностью правильным.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Некоторые замечания и рекомендации💡:</b> В случае, когда решение на отдельном шаге станет еще лучше, если внести небольшие коррективы. Желтые замечания и рекомендации необязательны к исправлению - на то это и рекомендации. Они больше ориентированы на улучшение твоих навыков
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>


<b>На доработку🤔:</b>
 В случае, когда решение на отдельном шаге требует существенной переработки и внесения правок. Напоминаю, что проект не может быть принят с первого раза, если ревью содержит комментарии, рекомендующие доработать шаги.
</div>

Увидев у тебя неточность, в первый раз я лишь укажу на ее наличие и дам тебе возможность самому найти и исправить ее. На реальной работе твой руководитель будет поступать также. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку!🤓

# Выбор локации для скважины

Допустим, вы работаете в добывающей компании «ГлавРосГосНефть». Нужно решить, где бурить новую скважину.

Вам предоставлены пробы нефти в трёх регионах: в каждом 10 000 месторождений, где измерили качество нефти и объём её запасов. Постройте модель машинного обучения, которая поможет определить регион, где добыча принесёт наибольшую прибыль. Проанализируйте возможную прибыль и риски техникой *Bootstrap.*

Шаги для выбора локации:

- В избранном регионе ищут месторождения, для каждого определяют значения признаков;
- Строят модель и оценивают объём запасов;
- Выбирают месторождения с самым высокими оценками значений. Количество месторождений зависит от бюджета компании и стоимости разработки одной скважины;
- Прибыль равна суммарной прибыли отобранных месторождений.

## Загрузка и подготовка данных

In [57]:
import pandas as pd
import numpy as np


from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import mean_squared_error, make_scorer
from scipy import stats as st

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Некоторые замечания и рекомендации💡:</b>

    
Хорошая структуризация импортов, но советовал бы не мешать раздел import ... и from ... import ...

<div class="alert alert-info">
    <h2> Комментарий студента <a class="tocSkip"> </h2>

    
Пересмотрел структуризацию импортов - сделал разделение по импорту целых библиотек и импорту методов из библиотек отдельно. Также удалил лишние импорты

<div class="alert alert-success">

<h2> Комментарий ревьюера 2 итерация <a class="tocSkip"> </h2>

Супер

In [2]:
data_r0 = pd.read_csv('/datasets/geo_data_0.csv')
data_r1 = pd.read_csv('/datasets/geo_data_1.csv')
data_r2 = pd.read_csv('/datasets/geo_data_2.csv')

Рассмотрим данные по регионам 0, 1 и 2 - посмотрим общую информацию, описание данных и первые пять строк по каждому региону.

In [3]:
display(data_r0.info())
display(data_r0.describe())
display(data_r0.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.500419       0.250143       2.502647      92.500000
std         0.871832       0.504433       3.248248      44.288691
min        -1.408605      -0.848218     -12.088328       0.000000
25%        -0.072580      -0.200881       0.287748      56.497507
50%         0.502360       0.250252       2.515969      91.849972
75%         1.073581       0.700646       4.715088     128.564089
max         2.362331       1.343769      16.003790     185.364347

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

In [4]:
display(data_r1.info())
display(data_r1.describe())
display(data_r1.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        1.141296      -4.796579       2.494541      68.825000
std         8.965932       5.119872       1.703572      45.944423
min       -31.609576     -26.358598      -0.018144       0.000000
25%        -6.298551      -8.267985       1.000021      26.953261
50%         1.153055      -4.813172       2.011479      57.085625
75%         8.621015      -1.332816       3.999904     107.813044
max        29.421755      18.734063       5.019721     137.945408

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

In [5]:
display(data_r2.info())
display(data_r2.describe())
display(data_r2.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.002023      -0.002081       2.495128      95.000000
std         1.732045       1.730417       3.473445      44.749921
min        -8.760004      -7.084020     -11.970335       0.000000
25%        -1.162288      -1.174820       0.130359      59.450441
50%         0.009424      -0.009482       2.484236      94.925613
75%         1.158535       1.163678       4.858794     130.595027
max         7.238262       7.844801      16.739402     190.029838

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

Судя по первичному анализу данных - данные были переданы предварительно подготовленные, пропусков нет, данные достаточно однородные.

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Некоторые замечания и рекомендации💡:</b>

Стоило бы посмотреть на распределения признаков, чтобы получить информацию о данных, с которыми стоит работать. Возможно, что где-то содержится ошибка и на это стоит сразу же обратить внимание с помощью анализа данных.
    
    
Можно было бы ещё провести корреляционный анализ, чтобы посмотреть на различия для наших нескольких датасетов - эта информация тоже может дать нам что-то. Плюс стоило бы обратить внимание на мультиколлиниарность, раз уж предстоит работать с линейными моделями

## Обучение и проверка модели

В качестве модели будем использовать линейную регрессию. Предвалительно подготовим данные для обучения и тестирования моделей - разделим данные для каждого региона на признаки и цели, после чего выделим выборки для кросс-валидации и обучения (CV-выборка) и тестовую выборку (test-выборка).

In [6]:
features_r0 = data_r0.drop(['id', 'product'], axis=1)
target_r0 = data_r0['product']

features_r1 = data_r1.drop(['id', 'product'], axis=1)
target_r1 = data_r1['product']

features_r2 = data_r2.drop(['id', 'product'], axis=1)
target_r2 = data_r2['product']

<div class="alert alert-info">
    <h2> Комментарий студента <a class="tocSkip"> </h2>

    
Тут я начал оборачивать часть кода в функции, что в итоге закончилось переписью проекта с использованием списка списков. Итоговый результат не поменялся, но было интересно попробовать реализовать проект по другому.

<div class="alert alert-success">

<h2> Комментарий ревьюера 2 итерация <a class="tocSkip"> </h2>

Отлично
    

Создадим список regions, в котором будут находиться все необходимые для расчётов данные в следующем формате: 

- regions[region_id][data_id]

где:
* region_id - порядковый номер региона, совпадающий с условинм задачи - 0, 1 и 2;
* data_id - id необходимых для рассчёта данных:
     * 0 - features_cv;
     * 1 - features_test;
     * 2 - target_cv;
     * 3 - target_test.

Список будет наполняться в ходе функции, проводящей разделение данных.\
Также для упрощения дальнейшей работы с данными target_test - в функции индексы этих данных будут обнулены, т.е. данные будут иметь новые индексы от 0 до 24999.

In [7]:
regions = []
def split(features, target):
    region_data = []
    features_cv, features_test, target_cv, target_test = train_test_split(features,
                                                                         target,
                                                                         test_size=0.25,
                                                                         shuffle=True,
                                                                         random_state=131296)
    region_data.append(features_cv)
    region_data.append(features_test)
    region_data.append(target_cv)
    target_test = target_test.reset_index(drop=True)
    region_data.append(target_test)
    return regions.append(region_data)

Производим сплит для всех регионов.

In [8]:
split(features_r0, target_r0)
split(features_r1, target_r1)
split(features_r2, target_r2)

Для обучения и проверки модели создадим функцию, которая будет проводить кросс-валидацию модели линейной регрессии на основании RMSE, обучать модель на CV-выборке, получать RMSE на test-выборке и рассчитывать среднее предсказанное количество нефти в тыс. баррелей на скважину. Все полученные значения функция будет выводить на печать.

Также в ходе работы функции для дальнейших рассчётов в список данных regions будут сохраняться списки с предсказаниями для каждого региона, которые будут переводится в формат pd.Series с новыми индексами от 0 до 24999, которые будут соответствовать записям из target_test.

Уточним список индексов в списке regions после добавления в него данных по предсказаниям:
* region_id - порядковый номер региона, совпадающий с условинм задачи - 0, 1 и 2;
* data_id - id необходимых для рассчёта данных:
     * 0 - features_cv;
     * 1 - features_test;
     * 2 - target_cv;
     * 3 - target_test;
     * 4 - predictions.

In [9]:
def model(region_id):
    model = LinearRegression()
    scores = cross_val_score(model, regions[region_id][0], regions[region_id][2], cv=5, scoring='neg_mean_squared_error')
    final_score = abs(np.mean(scores))
    model.fit(regions[region_id][0], regions[region_id][2])
    predictions = model.predict(regions[region_id][1])
    result = mean_squared_error(regions[region_id][3], predictions)**0.5
    predictions = pd.Series(predictions)
    
    return (regions[region_id].append(predictions),
            print('MSE на кросс-валидации:', final_score),
            print('RMSE на кросс-валидации:', final_score**0.5),
            print('RMSE на тестовой выборке:', result),
            print('Среднее предсказанное количество тыс. баррелей на скважину в регионе:', predictions.mean())
            )

In [10]:
for i in range(3):
    print(f'Регион {i}')
    model(i);
    print('-'*25)

Регион 0
MSE на кросс-валидации: 1417.5758099006212
RMSE на кросс-валидации: 37.65070796015157
RMSE на тестовой выборке: 37.82413746607025
Среднее предсказанное количество тыс. баррелей на скважину в регионе: 92.44406533876325
-------------------------
Регион 1
MSE на кросс-валидации: 0.7918412214864892
RMSE на кросс-валидации: 0.8898546069367114
RMSE на тестовой выборке: 0.8921725476909704
Среднее предсказанное количество тыс. баррелей на скважину в регионе: 68.88394552865586
-------------------------
Регион 2
MSE на кросс-валидации: 1605.69153579596
RMSE на кросс-валидации: 40.071081041019596
RMSE на тестовой выборке: 40.02150979913441
Среднее предсказанное количество тыс. баррелей на скважину в регионе: 95.0818059508208
-------------------------


<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Некоторые замечания и рекомендации💡:</b>

Тут возможно бы использовать функцию, чтобы не дублировать код. Если повторять его, то становится куда сложнее его поддерживать - приходится искать в коде изменяющиеся части и проводить замену несколько раз - вероятность ошибиться в таком случае куда выше, чем при использовании функций. Плюс чисто физически это занимает меньше места и становится проще читать

Самое маленькое значение RMSE на тестовой выборке было получени для Региона 1. При этом, в этом региона наименьшее среднее количество ресурсов на скважину по предсказанным данным.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 

Хорошо, согласен. Можно было бы ещё предложить методы по улучшению предсказания - собрать больше данных, выбрать другую модель и т.п., чтобы заказчику сразу располагал методами решения проблемы

## Подготовка к расчёту прибыли

Запишем в переменные необходимые данные из условия задачи, а также посчитаем некоторые дополнительные параметры для расчёта прибыли.

In [11]:
budget = 10000 #млн рублей
price = 0.45 #доходность 1 тыс баррелей в млн рублей
points_per_budget = 200 #количество точек разработки в одном бюджетировании

budget_per_point = budget/points_per_budget #бюджет в млн руб на 1 скважину
product_value_point = budget_per_point/price #минимальное необходимое количество добычи в тыс. баррелей для выхода скважины в 0

In [12]:
print('Бюджет всей разработки скважен (млн. руб.):', budget)
print('Доходность 1 тыс. баррелей (млн. руб.):', price)
print('Количество точке разработки скважин в рамках одного бюджета:', points_per_budget)
print('Бюджет на разработку одной скважины (млн. руб.):', budget_per_point)
print('Минимальная необходимая добычы в тыс. баррелей для получения прибыли со скважины:', product_value_point)

Бюджет всей разработки скважен (млн. руб.): 10000
Доходность 1 тыс. баррелей (млн. руб.): 0.45
Количество точке разработки скважин в рамках одного бюджета: 200
Бюджет на разработку одной скважины (млн. руб.): 50.0
Минимальная необходимая добычы в тыс. баррелей для получения прибыли со скважины: 111.11111111111111


In [13]:
print('Среднее количество тыс. баррелей на скважину в Регионе 0:', data_r0['product'].mean())
print('Среднее количество тыс. баррелей на скважину в Регионе 1:', data_r1['product'].mean())
print('Среднее количество тыс. баррелей на скважину в Регионе 2:', data_r2['product'].mean())

Среднее количество тыс. баррелей на скважину в Регионе 0: 92.50000000000001
Среднее количество тыс. баррелей на скважину в Регионе 1: 68.82500000000002
Среднее количество тыс. баррелей на скважину в Регионе 2: 95.00000000000004


Сравним необходимую минимальную добычу со скважины с предсказанными средними показателями по регионам:
* Регион 0 - среднее на скважину 92.50 < необходимо 111.11;
* Регион 1 - среднее на скважину 68.82 < необходимо 111.11;
* Регион 2 - среднее на скважину 95.00 < необходимо 111.11.

Все регионы несут риски, которые будем просчитывать далее. Для этого расчёта создадим функцию определения прибыли с некоторого количества скважин.

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>


<b>На доработку🤔:</b>

Средний объем необходимо считать, используя имеющиеся данные, а не предсказанные - вторые содержат ошибку и будет не совсем верно их представлять для объективного рассчета показателей.

<div class="alert alert-info">
    <h2> Комментарий студента <a class="tocSkip"> </h2>

    
Исправил - сначала посчитал среднее значение на скважину по генеральной совокупности по каждому региону, после этого сравнил необходимое для самоокупаемости с полученными значениями.

In [14]:
def revenue(target, predictions, count):
    predictions_sorted = predictions.sort_values(ascending=False)
    selected = target[predictions_sorted.index][:count]
    return price * selected.sum()

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>


<b>На доработку🤔:</b>

Необходимо использовать константы, которые были определены ранее. Без их использования код так же становиться сложнее поддерживать - необходимо менять параметры во всех частях, что используют переменную и это может повлечь за собой ошибки.
    
Так же можешь ознакомиться с рекомендациями по написанию кода на Python в целом
    
https://pythonworld.ru/osnovy/pep-8-rukovodstvo-po-napisaniyu-koda-na-python.html

<div class="alert alert-info">
    <h2> Комментарий студента <a class="tocSkip"> </h2>

    
Исправил - внес в функцию переменную price, созданную ранее

<div class="alert alert-success">

<h2> Комментарий ревьюера 2 итерация <a class="tocSkip"> </h2>

Плюс

## Расчёт прибыли и рисков 

Создадим функцию для рассчёта рисков с помощью бутстрепа. 

По условиям задачи, при разведке региона исследуют 500 точек, из которых выбирают 200 лучших. Для этого будем из общего списка истинных значений test-выборки отбирать семпл размером в 500 значений, получать предсказания с соответствующими индексами и с помощью созданной ранее функции получать прибыль 200 лучших скважин в семпле.

Список с прибылями по результатам бутстрепа обработаем с помощью цикла с целью получения количества испытаний, при которых добыча в регионе не окупилась - то есть потенциальная добыча с 200 скважин была меньше бюджета в 10 млрд. рублей.

По результатам работы функции получим среднюю выручку по бутстрепу, 95% доверительный интервал и риск убытка. Риск убытка будем считать как процент не окупившихся испытаний по отношению к значению цикла бутстрепа - 1000 испытаний.

In [15]:
def risk(region_id):
    state = np.random.RandomState(131296)
    values = []
 
    for i in range(1000):
        target_subsample = regions[region_id][3].sample(replace=True, random_state=state, n=500, axis=0)
        probs_subsample = regions[region_id][4][target_subsample.index]
        values.append(revenue(target_subsample, probs_subsample, 200))
 
    loss=0
    for i in values:
        if i < budget:
            loss += 1
        
    values = pd.Series(values)
    mean = values.mean()
    lower = values.quantile(0.025)
    upper = values.quantile(0.975)
 
    return (print("Средняя выручка:", mean),
            print("95%-ый доверительный интервал:", lower, upper),
            print("Риск убытков в процентах:", loss/1000*100))

In [16]:
for i in range(3):
    print(f'Регион {i}')
    risk(i);
    print('-'*25)

Регион 0
Средняя выручка: 10404.755982492421
95%-ый доверительный интервал: 9880.779295422853 10936.087505251335
Риск убытков в процентах: 6.7
-------------------------
Регион 1
Средняя выручка: 10490.767043045487
95%-ый доверительный интервал: 10109.508376367205 10924.478580219586
Риск убытков в процентах: 0.4
-------------------------
Регион 2
Средняя выручка: 10317.797155224289
95%-ый доверительный интервал: 9763.374050128548 10910.11832154286
Риск убытков в процентах: 12.6
-------------------------


<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Некоторые замечания и рекомендации💡:</b>

Такие большие дублирующие участки кода тоже стоило бы обернуть в функцию.

По результатам бутстрепа наиболее надёжным регионом является Регион 1. Риск добычи в этом регионе составляет всего 0.4 процента. Средняя выручка также в этом регионе больше, чем в остальных.

## Общий вывод

Задачей исследования был выбор наименее рискованного и наиболее прибыльного региона для разработки месторождений нефти. Выбор проводился среди трёх регионов, данные по которым были предоставлены.

Перед началом работы была проведена загрузка и первичный анализ данных по трём регионам. Данные были переданы уже подготовленные к анализу - пропуски отсутствуют, данные достаточно однородные.

Первым этапом исследования являлось обучение и проверки модели для каждого из региона. В качестве модели была использована линейная регрессия. Для каждого региона модель была проверена с помощью кросс-валидации с использованием требуемого скоринга - RMSE. После этого модели были обучены на CV-выборке данных и испытаны на test-выборке. Для каждой модели по регионам определены RMSE и среднее количество тыс. баррелей в скваженах по предсказаниям модели. Регионом с наименьшим RMSE был Регион 1, но при этом среднее количество тыс. баррелей в скважине в этом регионе наименьшее.

Вторым этапом исследования являлся сбор необходимых для дальнейших расчётов данных в переменные и расчёт дополнительных значений:
* Бюджет всей разработки скважен (млн. руб.): 10000
* Доходность 1 тыс. баррелей (млн. руб.): 0.45
* Количество точке разработки скважин в рамках одного бюджета: 200
* Бюджет на разработку одной скважины (млн. руб.): 50.0
* Минимальная необходимая добычы в тыс. баррелей для получения прибыли со скважины: 111.11

Также на этом этапе создаётся функция рассчёта прибыли лучших скважин с заданными истинными значениями, предсказаниями и количеством лучших скважин.

Третьим этапом исследования являлся рассчёт прибыли и рисков для добычи в каждом регионе. По условию исследования, при разведке региона исследуются 500 локаций для скважин, из них 200 лучших, согласно модели машинного обучения, разрабатываются. В ходе этого этапа проведена подготовка данных - списков истинных значений и предсказаний. Для списков предсказаний - изменён тип данных на pd.Series что создало для них новые индексы; для списков истинных значений - обнулены индексы. Это сделано с целью дальнейшего обращения по индексам соответствующих истинных значений и предсказаний. Был проведён бутстрепинг на 1000 испытаний, по результатам которого рассчитывались значения средней выручки, 95% доверительный интервал и процент риска. Процент риска рассчитывался как доля убыточных испытаний относительно общего количества испытаний.

По результатам исследования наиболее перспективным к разработки регионом можно считать Регион 1. При испытании с помощью бутстрепинга и модели машинного обучения результаты по Региону 1 следующие:
* Средняя выручка: 10490.767043045487
* 95%-ый доверительный интервал: 10109.508376367205 10924.478580219586
* Риск убытков в процентах: 0.4

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> 
    
Согласен с выводами, отлично, что пишешь ключевую информацию - это делает наш отчет более информативным. Чтобы ещё его улучшить можно добавить информацию по модели (метрики и т.п.), а так же идеи насчет улучшения результатов

<div style="border:solid Chocolate 2px; padding: 40px">

**Общий вывод по проекту (комментарий ревьера)**

Спасибо за твой проект! Есть несколько моментов, помеченных красным, которые стоит исправить. Если хочешь получить ещё больше опыта - попробуй воспользоваться частью советов из желтых комментариев. В целом, логика проекта верна, так что уверен, что тебе не составит труда исправить все остальное.

**Отмечу отдельные положительные моменты проекта🙂**:

* Достаточное комментирование по ходу работы
* Блоки кода структурированы и логичны

**На что стоит обратить внимание🤓**:

* Стоит переиспользовать код в виде функций и использовать константы
* Остальные рекомендации ты найдешь в моих комментариях!


**Желаю удачи и жду твой проект на повторное ревью!**😉


<div class="alert alert-info">
    <h2> Комментарий студента <a class="tocSkip"> </h2>

    
В итоге переделал проект с использованием новых функций и списка списков с данными по регионам. Единственное "желтое" замечание, которое я не отработал - корреляционный анализ данных, хотелось бы поподробнее узнать, как его можно провести примерно.

<div class="alert alert-success">

<h2> Комментарий ревьюера 2 итерация <a class="tocSkip"> </h2>

Можно посмотреть, например, здесь
    
https://habr.com/ru/post/557998/
    
https://medium.com/@dipti.rohan.pawar/correlation-statistical-analysis-9471411f0431

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: модели обучены и проверены
    - [x]  Данные корректно разбиты на обучающую и валидационную выборки
    - [x]  Модели обучены, предсказания сделаны
    - [x]  Предсказания и правильные ответы на валидационной выборке сохранены
    - [x]  На экране напечатаны результаты
    - [x]  Сделаны выводы
- [x]  Выполнен шаг 3: проведена подготовка к расчёту прибыли
    - [x]  Для всех ключевых значений созданы константы Python
    - [x]  Посчитано минимальное среднее количество продукта в месторождениях региона, достаточное для разработки
    - [x]  По предыдущему пункту сделаны выводы
    - [x]  Написана функция расчёта прибыли
- [x]  Выполнен шаг 4: посчитаны риски и прибыль
    - [x]  Проведена процедура *Bootstrap*
    - [x]  Все параметры бутстрепа соответствуют условию
    - [x]  Найдены все нужные величины
    - [x]  Предложен регион для разработки месторождения
    - [x]  Выбор региона обоснован

In [39]:
features_r1_cv, features_r1_test, target_r1_cv, target_r1_test = train_test_split(features_r1,
                                                                                  target_r1,
                                                                                  test_size=0.25,
                                                                                  shuffle=True,
                                                                                  random_state=131296)

In [51]:
def risk_score_2(y_true, y_pred):
    state = np.random.RandomState(131296)
    values = []
    
    y_true = y_true.reset_index(drop=True)
    y_pred = pd.Series(y_pred)
    
    for i in range(1000):
        target_subsample = y_true.sample(replace=True, random_state=state, n=500, axis=0)
        probs_subsample = y_pred[target_subsample.index]
        values.append(revenue(target_subsample, probs_subsample, 200))
 
    loss=0
    for i in values:
        if i < budget:
            loss += 1
    
    error = loss/10
    return error

In [52]:
risk_scorer_2 = make_scorer(risk_score_2, greater_is_better=False)

In [53]:
model_test = LinearRegression()
scores = cross_val_score(model_test, features_r1_cv, target_r1_cv, cv=5, scoring=risk_scorer_2)

In [56]:
print(scores)
print(abs(np.mean(scores)))

[-0.5 -0.4 -0.7 -0.2 -0.4]
0.44000000000000006
